action/protocol/staking/candidate_statemanager.go

```diff
@@ -22,9 +27,28 @@ const (
 )
 
 type (
+	// BucketSet related to setting bucket
+	BucketSet interface {
+		updateBucket(index uint64, bucket *VoteBucket) error
+		putBucket(bucket *VoteBucket) (uint64, error)
+		delBucket(index uint64) error
+		putBucketAndIndex(bucket *VoteBucket) (uint64, error)
+		delBucketAndIndex(owner, cand address.Address, index uint64) error
+		putBucketIndex(key []byte, index uint64) error
+		delBucketIndex(key []byte, index uint64) error
+	}
+	// CandidateSet related to setting candidates
+	CandidateSet interface {
+		delCandidate(name address.Address) error
+		putCandidate(d *Candidate) error
+	}
 	// CandidateStateManager is candidate state manager on top of StateManager
 	CandidateStateManager interface {
 		protocol.StateManager
```

Comment:
#2142 wants CandidateStateManager not to expose StateManager interface 
so we need to implement methods like PutState(), DelState() for (*candSM)
then we can delete "protocol.StateManager" here
@CoderZhi  can confirm

Reply:
should look like:
candSM struct {
		sm protocol.StateManager
		candCenter *CandidateCenter
		bucketPool *BucketPool
	}

and 
func (csm *candSM) PutState(s interface{}, options ...StateOption) (uint64, error) {
    return csm.sm.PutState(s, options...)
}

Reply:
After hide `StateManager ` interface , I think we may not modify `candSM` structure. If `candSM` is not changed , it has `StateManager` interface , so there is no need to implement functions like `PutState()`, `DelState()` for `candSM`.The code will be more simplified.

Reply:
just like `candSR` in candidate_statereader.go